test/harness.test.js

```diff
@@ -0,0 +1,170 @@
+"use strict";
+
+const { readFileSync, promises: fs } = require("fs");
+const path = require("path");
+const vm = require("vm");
+
+process.on("unhandledRejection", e => {
+  throw e;
+});
+
+function createSandbox(setupScripts) {
+  const sandbox = vm.createContext();
+  sandbox.self = vm.runInContext("this", sandbox);
+  for (const script of setupScripts) {
+    script.runInContext(sandbox);
+  }
+  sandbox.require = p => {
+    if (p[0] !== ".") {
+      // eslint-disable-next-line global-require
+      return require(p);
+    }
+    if (p.startsWith("../implementations/")) {
+      return { implementation: class Stub {} };
+    }
+
+    const resolved = path.resolve(__dirname, "./output", p);
+    const src = readFileSync(resolved, { encoding: "utf8" });
+    return vm.runInContext(
+      `(() => {
+        let exports = {};
+        const module = { exports };
+        (() => {${src}})();
+        return module.exports;
+      })();`,
+      sandbox,
+      { filename: resolved, lineOffset: 5, columnOffset: 16 }
+    );
```

Comment:
It might be better to use [`vm.compileFunction(…)`](https://nodejs.org/api/vm.html#vm_vm_compilefunction_code_params_options) here.